src/Microsoft.AspNetCore.Blazor.Server/Circuits/DefaultCircuitFactory.cs

```diff
@@ -0,0 +1,58 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using Microsoft.AspNetCore.Blazor.Browser.Rendering;
+using Microsoft.AspNetCore.Blazor.Rendering;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.SignalR;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Microsoft.AspNetCore.Blazor.Server.Circuits
+{
+    internal class DefaultCircuitFactory : CircuitFactory
+    {
+        private readonly IServiceScopeFactory _scopeFactory;
+
+        public DefaultCircuitFactory(IServiceScopeFactory scopeFactory)
+        {
+            _scopeFactory = scopeFactory ?? throw new ArgumentNullException(nameof(scopeFactory));
+
+            StartupActions = new List<(PathString path, Action<BrowserRenderer> config)>();
```

Comment:
This looks good. It's nice to see this solution for how to configure multiple server-side Blazor apps while only having one top-level service provider.